protocol/communities/manager.go

```diff
@@ -991,38 +992,81 @@ func (m *Manager) EditCommunityTokenPermission(request *requests.EditCommunityTo
 	return community, changes, nil
 }
 
+type reevaluateMemberRole struct {
+	old protobuf.CommunityMember_Roles
+	new protobuf.CommunityMember_Roles
+}
+
+func (rmr reevaluateMemberRole) hasChanged() bool {
+	return rmr.old != rmr.new
+}
+
+func (rmr reevaluateMemberRole) isPrivileged() bool {
+	return rmr.new != protobuf.CommunityMember_ROLE_NONE
+}
+
+func (rmr reevaluateMemberRole) hasChangedToPrivileged() bool {
+	return rmr.hasChanged() && rmr.old == protobuf.CommunityMember_ROLE_NONE
+}
+
+type reevaluateMembersResult struct {
+	membersToRemove             map[string]struct{}
+	membersRoles                map[string]*reevaluateMemberRole
+	membersToRemoveFromChannels map[string]map[string]struct{}
+	membersToAddToChannels      map[string]map[string]protobuf.CommunityMember_ChannelRole
+}
+
+func (rmr *reevaluateMembersResult) newPrivilegedRoles() (map[protobuf.CommunityMember_Roles][]*ecdsa.PublicKey, error) {
+	result := map[protobuf.CommunityMember_Roles][]*ecdsa.PublicKey{}
+
+	for memberKey, roles := range rmr.membersRoles {
+		if roles.hasChangedToPrivileged() {
+			memberPubKey, err := common.HexToPubkey(memberKey)
+			if err != nil {
+				return nil, err
+			}
+			if result[roles.new] == nil {
+				result[roles.new] = []*ecdsa.PublicKey{}
+			}
+			result[roles.new] = append(result[roles.new], memberPubKey)
+		}
+	}
+
+	return result, nil
+}
+
 // use it only for testing purposes
 func (m *Manager) ReevaluateMembers(communityID types.HexBytes) (*Community, map[protobuf.CommunityMember_Roles][]*ecdsa.PublicKey, error) {
 	return m.reevaluateMembers(communityID)
 }
 
+// First, the community is read from the database,
+// then the members are reevaluated, and only then
+// the community is locked and changes are applied.
+// NOTE: Changes made to the same community
+// while reevaluation is ongoing are respected
+// and do not affect the result of this function.
+// If permissions are changed in the meantime,
+// they will be accommodated with the next reevaluation.
 func (m *Manager) reevaluateMembers(communityID types.HexBytes) (*Community, map[protobuf.CommunityMember_Roles][]*ecdsa.PublicKey, error) {
-	m.communityLock.Lock(communityID)
-	defer m.communityLock.Unlock(communityID)
-
```

Comment:
Should you lock to get the community just in case it is being modified by something at the same time or it's should be a non-issue?

Reply:
Good point but I consider this case negligible. Even if that happens, the next reevaluation will take these changes into consideration.